fix(search): prevent strpos(unknown) error on PostgreSQL when searching link fields without translation

dhiashalabi opened the pull request:
<!-- no-docs -->

### **Description**

This PR fixes an issue in PostgreSQL query transformation where `LOCATE(needle, haystack[, start])` was rewritten incorrectly to `STRPOS(...)` when the `needle` contained a comma (`,`) or other special characters inside a quoted string (e.g. `locate(',', name)`). This could result in invalid SQL and failures in APIs that rely on `LOCATE()`, including the Link field search API (`frappe.desk.search.search_link`).

The root cause was the previous **regex-based** `LOCATE()` rewrite in the PostgreSQL query modifier, which could not reliably split function arguments when commas appeared inside quoted literals.

This change replaces the regex substitution with a small, safe argument parser that:

* correctly splits `LOCATE()` arguments while respecting quoted strings and nested expressions
* supports both **2-argument** and **3-argument** `LOCATE()` forms
* **fails closed** (leaves the query unchanged) if parsing is uncertain, avoiding generation of broken SQL

---

### **Before**

* On PostgreSQL, queries containing `LOCATE(',', ...)` could be transformed incorrectly.
* Link field searches and other LOCATE-based queries could fail with database errors.

### **After**

* `LOCATE(needle, haystack[, start])` is rewritten reliably on PostgreSQL, even when the needle contains commas or special characters.
* Link field search and other LOCATE-based queries work as expected without producing invalid SQL.

---

**Closes:** #34533

dhiashalabi committed:
fix(search): prevent strpos(unknown) error on PostgreSQL for un-translated doctypes

dhiashalabi commented:
### Backport Request

This fix addresses a PostgreSQL query transformation bug where `LOCATE(needle, haystack[, start])` could be rewritten incorrectly to `STRPOS(...)` when the *needle* contains a comma (`,`), producing invalid SQL and causing Link field search (`frappe.desk.search.search_link`) and other LOCATE-based queries to fail on PostgreSQL.

Since this can affect production sites on both v14 and v15 running PostgreSQL, could this patch please be backported to:

* version-15
* version-14

akhilnarang commented:
@dhiashalabi could you see if you can fix this by adjusting this regex instead?

- https://github.com/frappe/frappe/blob/develop/frappe/database/postgres/database.py#L499
- https://github.com/frappe/frappe/blob/develop/frappe/database/postgres/database.py#L544

(I haven't checked how strpos needs the arguments, maybe we can add `cast as varchar` but if it could be resolved here itself, would make more sense, if not possible, can go ahead with this).

dhiashalabi commented:
> @dhiashalabi could you see if you can fix this by adjusting this regex instead?
> 
> * https://github.com/frappe/frappe/blob/develop/frappe/database/postgres/database.py#L499
> * https://github.com/frappe/frappe/blob/develop/frappe/database/postgres/database.py#L544
> 
> (I haven't checked how strpos needs the arguments, maybe we can add `cast as varchar` but if it could be resolved here itself, would make more sense, if not possible, can go ahead with this).

@akhilnarang Thank you for the feedback. I investigated fixing this in the regex, but it's not feasible.

**The Problem:**
The regex `locate\(([^,]+),([^)]+)(\)?)\)` fails when:
1. Search text contains commas: `locate(',', field)` 
2. Arguments have nested parentheses: `locate(text, CAST(field AS varchar))`
3. Regex can't track parenthesis depth or quote context

**Why regex-only fails:**
SQL requires stateful parsing for nested structures. To handle all cases, we'd need ~70 lines of parser code with significant risk for edge cases across the entire codebase.

akhilnarang commented:
Got it @dhiashalabi. Since we already have some if-elif checks for mariadb and postgres just a few lines below - could you move these changes in there as well?

dhiashalabi committed:
Merge branch 'develop' into fix/postgres-strpos-error

dhiashalabi committed:
fix(search_link): use Postgres-safe relevance expr; keep QB expr for mariadb/sqlite

dhiashalabi committed:
revert(search): restore develop search.py (handled in Postgres query transformer)

dhiashalabi committed:
fix(postgres): robust LOCATE() rewrite to STRPOS() in query transformer

dhiashalabi committed:
fix(postgres): robust LOCATE() -> STRPOS() rewrite (handles commas in strings)

dhiashalabi commented:
Hey @akhilnarang — yes, I agree it makes more sense to fix this at the Postgres query-transformer level rather than in `search.py`.

The root issue isn’t `strpos()` casting, it’s that the current regex-based `LOCATE(...)` rewrite breaks when the *needle* contains a comma inside a quoted literal (e.g. `locate(',', name)`), so the substitution produces malformed SQL and triggers the Postgres error.

I’ve updated `frappe/database/postgres/database.py` to replace the regex substitution with a small, safe parser that:

* correctly splits `LOCATE(needle, haystack[, start])` args while respecting quoted strings + nested expressions
* supports both 2-arg and 3-arg `LOCATE`
* fails closed (leaves the query unchanged) if parsing is uncertain, avoiding broken SQL

With this in place, I reverted the earlier `search.py` workaround and kept the existing mariadb/postgres branching behavior there unchanged.

ankush commented on frappe/database/postgres/database.py:
```diff
@@ -538,8 +538,164 @@ def modify_value(value):
-	if LOCATE_QUERY_PATTERN.search(query):
-		query = LOCATE_SUB_PATTERN.sub(r"strpos(\2\3, \1)", query)
-	return query
+def replace_locate_with_strpos(query: str) -> str:
```
This much magical query rewrite doesn't make sense IMO. 

We should just expect queries to be rewritten where `locate` is used? @AarDG10

ankush replied:
The only non-magical way to do this would be to add support for it in query builder and handle all DB specific transformations there. 


Queries will HAVE to be rewritten, especially stuff like `search.py`.
